presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/IntermediateAggregationOptimizer.java

```diff
@@ -0,0 +1,335 @@
+        {
+            return context.defaultRewrite(node, Optional.empty(), this::retainFirstSource);
+        }
+
+        private Boolean retainFirstSource(List<Boolean> booleans)
+        {
+            return booleans.get(0);
+        }
+
+        private Boolean retainIfAny(List<Boolean> booleans)
+        {
+            return booleans.stream().anyMatch(isSourcePartitioned -> isSourcePartitioned);
+        }
+    }
+
+    private static class AggregationSpecification
+    {
+        private final Map<Symbol, AggregationNode.Aggregation> assignments;
+        private final List<List<Symbol>> groupingSets;
+        private final Optional<Symbol> hashSymbol;
+        private final Optional<Symbol> groupIdSymbol;
+
+        public AggregationSpecification(
+                Map<Symbol, AggregationNode.Aggregation> assignments,
+                List<List<Symbol>> groupingSets,
+                Optional<Symbol> hashSymbol,
+                Optional<Symbol> groupIdSymbol)
+        {
+            this.assignments = ImmutableMap.copyOf(requireNonNull(assignments, "assignments is null"));
+            this.groupingSets = ImmutableList.copyOf(requireNonNull(groupingSets, "groupingSets is null"));
+            this.hashSymbol = requireNonNull(hashSymbol, "hashSymbol is null");
+            this.groupIdSymbol = requireNonNull(groupIdSymbol, "groupIdSymbol is null");
+        }
+
+        public static AggregationSpecification extract(AggregationNode node)
+        {
+            // Rewrite assignments to output the same symbol as the input. This allows the generated aggregation
+            // to be easily inserted anywhere in the plan with a simple translation.
+            ImmutableMap.Builder<Symbol, AggregationNode.Aggregation> assignments = ImmutableMap.builder();
+            for (Map.Entry<Symbol, AggregationNode.Aggregation> entry : node.getAssignments().entrySet()) {
+                // There should only be one input symbol for aggregation functions
+                Symbol inputSymbol = Iterables.getOnlyElement(DependencyExtractor.extractAll(entry.getValue().getCall()));
+                assignments.put(inputSymbol, entry.getValue());
+            }
+
+            return new AggregationSpecification(
+                    assignments.build(),
+                    node.getGroupingSets(),
+                    node.getHashSymbol(),
+                    node.getGroupIdSymbol());
+        }
+
+        public AggregationNode applyAfter(PlanNodeIdAllocator idAllocator, AggregationNode.Step step, PlanNode source)
+        {
+            return new AggregationNode(
+                    idAllocator.getNextId(),
+                    source,
+                    assignments,
+                    groupingSets,
+                    step,
+                    hashSymbol,
+                    groupIdSymbol);
+        }
+
+        public AggregationSpecification translate(Function<Symbol, Symbol> translator)
```

Comment:
`SymbolMapper` is already in master branch, so you should be able to find it after rebasing your branch.